towhee/dataframe/iterators.py

```diff
@@ -0,0 +1,199 @@
+        Raises:
+            (`StopIteration`)
+                The iteration end if the `DataFrame` is sealed and the last row is
+                reached.
+        """
+        if self._done:
+            raise StopIteration
+
+        df = self._df_ref()
+
+        code, row = df.get(self._offset, count = self._batch_size, iter_id = self._id)
+
+        if code == Responses.INDEX_GC:
+            raise IndexError
+
+        elif code == Responses.INDEX_OOB_UNSEALED:
+            if self._block:
+                df.notify_map_block(self._id, self._event, self._offset, self._batch_size)
+                self._event.wait()
+                self._event.clear()
+                return self.__next__()
+
+            return None
+
+        elif code == Responses.APPROVED_CONTINUE:
+            # subtract one due to step.
+            df.ack(self._id, self._offset + self._step - 1)
+            self._offset += self._step
+            return row
+
+        elif code == Responses.INDEX_OOB_SEALED:
+            raise StopIteration
+
+        elif code == Responses.APPROVED_DONE:
+            self._done = True
+            df.ack(self._id, float('inf'))
+            self._offset = 0
+            return row
+
+        elif code == Responses.KILLED:
+            raise StopIteration
+
+        else: # 'unkown_error'
+            raise Exception
+
+    @property
+    def id(self):
+        return self._id
+
+
+class MapIterator(BaseIterator):
+    """
+    A row-based map `DataFrame` iterator.
+
+    Args:
+        df (`towhee.Dataframe`):
+            The dataframe that is being iterated.
+        block (`bool`):
+            Whether to block when data not present.
+    """
+    def __init__(self, df: DataFrame, block = True):
+        super().__init__(df, batch_size = 1, step = 1, block = block)
+
+
+class BatchIterator(BaseIterator):
+    """
+    A row-based batched map `DataFrame` iterator.
+
+    Args:
+        df (`towhee.Dataframe`):
+            The dataframe that is being iterated.
+        batch_size (`int`):
+            How many values to read.
+        step (`int`):
+            How many steps to take after each read.
+        block (`bool`):
+            Whether to block when data not present.
+    """
+    def __init__(self, df: DataFrame, batch_size = 1, step = 1, block = True):
+        super().__init__(df, batch_size = batch_size, step = step, block = block)
+
+
+class WindowIterator(BaseIterator):
+    """
+    A row-based window `DataFrame` iterator.
+
+    Args:
+        df (`towhee.Dataframe`):
+            The dataframe that is being iterated.
+        window_size (`int`):
+            How large of a window.
+        use_timestamp (`bool`):
+            Whether to use timestamp instead of row_id.
+        block (`bool`):
+            Whether to block when data not present.
+    """
+    def __init__(self, df: DataFrame, window_size = 1, use_timestamp = False, block = True):
```

Comment:
TimeWindowIterator's init maybe like this
```python
     def __init__(self, df: DataFrame, time_range_sec: int, time_step_sec: int, block=False):
        '''
        time_range_sec: window range
        time_step_sec: window step
        '''
        pass
```

Reply:
sounds good. There might be some complexity added but it shouldnt be too bad.